File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        int stair[] = new int[t];
        for (int i = 0; i < t; i++) {
            stair[i] = Integer.parseInt(br.readLine());
        }
        int DP[] = new int[t];
        int ans = 0;
        if (t <= 2) {
            for (int i = 0; i < t; i++) {
                ans += stair[i];
            }
        } else {
            DP[0] = stair[0];
            DP[1] = stair[0] + stair[1];
            DP[2] = Math.max(stair[0] + stair[2], stair[1] + stair[2]);
            for (int i = 3; i < t; i++) {
                DP[i] = Math.max(DP[i-3]+stair[i-1]+stair[i], DP[i-2]+stair[i]);
            }
            ans = DP[t-1];
        }
        System.out.println(ans);
    }
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys
input = sys.stdin.readline


n = int(input())
stair = [int(input()) for _ in range(n)]
DP = [0]*n

if n <= 2:
    print(sum(stair))

else:
    DP[0] = stair[0]
    DP[1] = stair[0] + stair[1]
    for i in range(2,n):
        DP[i]=max(DP[i-3]+stair[i-1]+stair[i], DP[i-2]+stair[i])
    print(DP[-1])